goagen/gen_controller/generator.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	return g
+}
+
+// Generator is the application code generator.
+type Generator struct {
+	API       *design.APIDefinition // The API definition
+	OutDir    string                // Path to output directory
+	DesignPkg string                // Path to design package, only used to mark generated files.
+	AppPkg    string                // Name of generated "app" package
+	Force     bool                  // Whether to override existing files
+	Pkg       string                // Name of the generated package
+	Resource  string                // Name of the generated file
+	genfiles  []string              // Generated files
+}
+
+// Generate is the generator entry point called by the meta generator.
+func Generate() (files []string, err error) {
+	var (
+		outDir, designPkg, appPkg, ver, res, pkg string
+		force                                    bool
+	)
+
+	set := flag.NewFlagSet("controller", flag.PanicOnError)
+	set.StringVar(&outDir, "out", "", "")
+	set.StringVar(&designPkg, "design", "", "")
+	set.StringVar(&appPkg, "app-pkg", "app", "")
+	set.StringVar(&pkg, "pkg", "controller", "")
+	set.StringVar(&res, "res", "", "")
+	set.StringVar(&ver, "version", "", "")
+	set.BoolVar(&force, "force", false, "")
+	set.Bool("notest", false, "")
+	set.Parse(os.Args[1:])
+
+	if err := codegen.CheckVersion(ver); err != nil {
+		return nil, err
+	}
+
+	appPkg = codegen.Goify(appPkg, false)
+	g := &Generator{OutDir: outDir, DesignPkg: designPkg, AppPkg: appPkg, Force: force, API: design.Design, Pkg: pkg, Resource: res}
+
+	return g.Generate()
+}
+
+// Generate produces the skeleton controller service factory.
+func (g *Generator) Generate() (_ []string, err error) {
+	go utils.Catch(nil, func() { g.Cleanup() })
+
+	defer func() {
+		if err != nil {
+			g.Cleanup()
+		}
+	}()
+
+	if g.AppPkg == "" {
+		g.AppPkg = "app"
+	}
+
+	codegen.Reserved[g.AppPkg] = true
+
+	funcs := template.FuncMap{
+		"tempvar":   tempvar,
+		"okResp":    g.okResp,
+		"targetPkg": func() string { return g.AppPkg },
+	}
+
+	imp, err := codegen.PackagePath(g.OutDir)
+	if err != nil {
+		return nil, err
+	}
+	imp = path.Join(filepath.ToSlash(imp), "app")
+
+	imports := []*codegen.ImportSpec{
+		codegen.SimpleImport("io"),
+		codegen.SimpleImport("github.com/goadesign/goa"),
+		codegen.SimpleImport(imp),
+		codegen.SimpleImport("golang.org/x/net/websocket"),
+	}
+
+	err = g.API.IterateResources(func(r *design.ResourceDefinition) error {
+		filename := filepath.Join(g.OutDir, codegen.SnakeCase(r.Name)+".go")
+
+		if g.Resource != "" {
+			if g.Resource == r.Name {
+				filename = filepath.Join(g.OutDir, codegen.SnakeCase(g.Resource)+".go")
+				g.generateControllerFile(filename, funcs, imports, r)
+			}
+		} else {
+			g.generateControllerFile(filename, funcs, imports, r)
+		}
+
+		return nil
+	})
+
+	return g.genfiles, nil
+}
+
+func (g *Generator) generateControllerFile(filename string, funcs template.FuncMap, imports []*codegen.ImportSpec, r *design.ResourceDefinition) error {
```

Comment:
Please refactor the `genmain` package slightly so that it exposes the function that generates the code (i.e. make a function out of https://github.com/goadesign/goa/blob/master/goagen/gen_main/generator.go#L112-L139) for a single controller. call it from here and remove all the copy/pasted code.